src/libaktualizr/uptane/managedsecondary.cc

```diff
@@ -23,11 +23,23 @@ ManagedSecondary::ManagedSecondary(const SecondaryConfig &sconfig_in) : Secondar
   public_key_id = Crypto::getKeyId(public_key);
 }
 
-bool ManagedSecondary::putMetadata(const MetaPack &meta_pack) {
+void ManagedSecondary::rawToMeta() {
+  // raw meta is trusted
+  current_meta.director_root = Uptane::Root("director", Utils::parseJSON(current_raw_meta.director_root));
+  current_meta.director_targets = Uptane::Targets(Utils::parseJSON(current_raw_meta.director_targets));
+  current_meta.image_root = Uptane::Root("repo", Utils::parseJSON(current_raw_meta.image_root));
+  current_meta.image_targets = Uptane::Targets(Utils::parseJSON(current_raw_meta.image_targets));
+  current_meta.image_timestamp = Uptane::TimestampMeta(Utils::parseJSON(current_raw_meta.image_timestamp));
+  current_meta.image_snapshot = Uptane::Snapshot(Utils::parseJSON(current_raw_meta.image_snapshot));
+}
+
+bool ManagedSecondary::putMetadata(const RawMetaPack &meta_pack) {
   // No verification is currently performed, we can add verification in future for testing purposes
   detected_attack = "";
-  current_meta = meta_pack;
-  storeMetadata(current_meta);
+
+  current_raw_meta = meta_pack;
+  rawToMeta();
+  storeMetadata(current_raw_meta);
```

Comment:
We don't need to be able to store a full metapack all at once. Each piece of metadata should be loaded one at at time so it's easier to pick up where an error happened. Imagine the case where we get a metapack with `snapshot` version `X`. This points to some `x_1, x_2, x_3, ...` versions of delegated metadata. If we can't load the metapack, but we get a snapshot later with version `X+1`, but versions of the delegated metadata don't match, this should be an error, but our case won't catch it (I think).

Reply:
This shouldn't happen because the metapack is stored/loaded atomically. So if we're still on snapshot version X, we'll be using the old metapack. When we download the new snapshot.X+1, we also download the whole metadata set, cross-check it and only if it's valid update the metadata. Doesn't it work?

Reply:
Not quite. If the new metapack has a bad `targets` metadata, the correct thing to do is still to update everything else.